File: src/main/java/dev/vxrp/util/colors/ColorTool.java
package dev.vxrp.util.colors;

import dev.vxrp.util.Enums.DCColor;
import org.jetbrains.annotations.NotNull;

public class ColorTool {

    private static final String filler = "\u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E " +
            "\u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E " +
            "\u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E " +
            "\u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E " +
            "\u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E " +
            "\u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E \u200E " +
            "\u200E \u200E \u200E \u200E \u200E \u200E \u200E";
    public static final String DARK_GRAY = "\u001B[2;30m";
    public static final String RED = "\u001B[31m";
    public static final String GREEN = "\u001B[32m";
    public static final String GOLD = "\u001B[33m";
    public static final String LIGHT_BLUE = "\u001B[34m";
    public static final String PINK = "\u001B[35m";
    public static final String TEAL = "\u001B[36m";
    public static final String WHITE = "\u001B[37m";
    public static final String BOLD = "\u001B[1;2m";
    public static final String UNDERLINE = "\u001B[4;2m";
    public static final String RESET = "\u001B[0m";


    @NotNull
    public static String apply(DCColor color, String text) {
        switch (color) {
            case DARK_GRAY -> {
                return "\u001B[2;30m"+text+"\u001B[0m";
            }
            case RED -> {
                return "\u001B[2;31m"+text+"\u001B[0m";
            }
            case GREEN -> {
                return "\u001B[2;32m"+text+"\u001B[0m";
            }
            case GOLD -> {
                return "\u001B[2;33m"+text+"\u001B[0m";
            }
            case LIGHT_BLUE -> {
                return "\u001B[2;34m"+text+"\u001B[0m";
            }
            case PINK -> {
                return "\u001B[2;35m"+text+"\u001B[0m";
            }
            case TEAL -> {
                return "\u001B[2;36m"+text+"\u001B[0m";
            }
            case WHITE -> {
                return "\u001B[2;37m"+text+"\u001B[0m";
            }
            case BOLD -> {
                return "\u001B[1;2m"+text+"\u001B[0m";
            }
            case UNDERLINE -> {
                return "\u001B[4;2m"+text+"\u001B[0m";
            }
        }
        return text;
    }
    @NotNull
    public static String useCustomColorCodes(String text) {
        return text
                .replace("&dark_gray&", "\u001B[2;30m")
                .replace("&red&", "\u001B[2;31m")
                .replace("&green&", "\u001B[2;32m")
                .replace("&gold&", "\u001B[2;33m")
                .replace("&light_blue&", "\u001B[2;34m")
                .replace("&pink&", "\u001B[2;35m")
                .replace("&teal&", "\u001B[2;36m")
                .replace("&white&", "\u001B[2;37m")
                .replace("&bold&", "\u001B[1;2m")
                .replace("&reset&", "\u001B[0m")
                .replace("&underline&", "\u001B[4;2m")
                .replace("%filler%", filler);

    }
    @NotNull
    public static String replaceCustomColorCode(String colorCode) {
        switch (colorCode) {
            case "&dark_gray&" -> {
                return "\u001B[2;30m";
            }
            case "&red&" -> {
                return "\u001B[2;31m";
            }
            case "&green&" -> {
                return "\u001B[2;32m";
            }
            case "&gold&" -> {
                return "\u001B[2;33m";
            }
            case "&light_blue&" -> {
                return "\u001B[2;34m";
            }
            case "&pink&" -> {
                return "\u001B[2;35m";
            }
            case "&teal&" -> {
                return "\u001B[2;36m";
            }
            case "&white&" -> {
                return "\u001B[2;37m";
            }
            case "&bold&" -> {
                return "\u001B[1;2m";
            }
            case "&reset&" -> {
                return "\u001B[0m";
            }
            case "&underline&" -> {
                return "\u001B[4;2m";
            }
        }
        return colorCode;
    }
}

File: src/main/kotlin/dev/vxrp/util/color/ColorTool.kt
/*
 * Copyright (c) 2024 Vxrpenter and the SCPToolsBot Contributors
 *
 * Licenced under the MIT License, any non-license compliant usage of this file(s) content
 * is prohibited. If you did not receive a copy of the license with this file, you
 *  may obtain the license at
 *
 *  https://mit-license.org/
 *
 *  This software may be used commercially if the usage is license compliant. The software
 *  is provided without any sort of WARRANTY, and the authors cannot be held liable for
 *  any form of claim, damages or other liabilities.
 *
 *  Note: This is no legal advice, please read the license conditions
 */

package dev.vxrp.util.color

import dev.vxrp.util.color.enums.DCColor
import dev.vxrp.util.color.enums.DCColor.*
import java.util.regex.Pattern
import kotlin.math.max
import kotlin.math.roundToInt

class ColorTool {
    private val singleFiller: String = "\u200E "

    private val darkGray: String = "\u001B[2;30m"
    private val red: String = "\u001B[31m"
    private val green: String = "\u001B[32m"
    private val gold: String = "\u001B[33m"
    private val lightBlue: String = "\u001B[34m"
    private val pink: String = "\u001B[35m"
    private val teal: String = "\u001B[36m"
    private val white: String = "\u001B[37m"
    private val bold: String = "\u001B[1;2m"
    private val underline: String = "\u001B[4;2m"
    private val reset: String = "\u001B[0m"

    fun apply(color: DCColor, text: String): String {
        return when (color) {
            DARK_GRAY -> "$darkGray$text$reset"
            RED -> "$red$text$reset"
            GREEN -> "$green$text$reset"
            GOLD -> "$gold$text$reset"
            LIGHT_BLUE -> "$lightBlue$text$reset"
            PINK -> "$pink$text$reset"
            TEAL -> "$teal$text$reset"
            WHITE -> "$white$text$reset"
            BOLD -> "$bold$text$reset"
            UNDERLINE -> "$underline$text$reset"
        }
    }

    fun parse(text: String): String {
        val matcher = Pattern.compile("(?<=&filler<).+?(?=>&|$)").matcher(text)
        while (matcher.find()) {
            val count = (matcher.group().toInt().toFloat() / 2).roundToInt()
            return text.replace("&filler<" + matcher.group() + ">&", singleFiller.repeat(max(0.0, count.toDouble()).toInt()))
        }

        return text
            .replace("&dark_gray&", darkGray)
            .replace("&red&", red)
            .replace("&green&", green)
            .replace("&gold&", gold)
            .replace("&light_blue&", lightBlue)
            .replace("&pink&", pink)
            .replace("&teal&", teal)
            .replace("&white&", white)
            .replace("&bold&", bold)
            .replace("&reset&", reset)
            .replace("&underline&", underline)
            .replace("&filler&", singleFiller.repeat(max(0.0, 144.toDouble()).toInt()))
            .replace("&singleFiller&", singleFiller)
    }
}